c4.pl

```diff
@@ -35,19 +46,117 @@
+% Cumulative Distribution: 0.5erfc(0.589256(4-x))
+% 7 appears 2% of time
+% 6 appears 9% of time
+% 5 appears 23% of time
+% 4 appears 32% of time
+% 3 appears 23% of time
+% 2 appears 9% of time
+% 1 appears 2% of time
+computerMove(_, normalDistribution, [Move], _, Move).
+computerMove(_, normalDistribution, AvailableMoves, _, Move) :-
+    length(AvailableMoves, L),
+    L > 1,
+    random_between(0, 100, X),
+    getMoveFromDistribution(X, MoveAttempt),
+    (notMember(MoveAttempt, AvailableMoves) -> 
+    computerMove(_, normalDistribution, AvailableMoves, _, Move);
+    Move is MoveAttempt).
+
+getMoveFromDistribution(X, Move) :-
+    ((X < 2) -> Move is 1 ;
+    (2 =< X, X < 11) -> Move is 2 ;
+    (11 =< X, X < 34) -> Move is 3 ;
+    (34 =< X, X < 66) -> Move is 4 ;
+    (66 =< X, X < 89) -> Move is 5 ;
+    (89 =< X, X < 98) -> Move is 6 ;
+    Move is 7).
+
+
+% Minimax AI with certain depth
+% Code inspired from https://www.metalevel.at/conn4/ and https://en.wikipedia.org/wiki/Alpha%E2%80%93beta_pruning#Pseudocode and modified to fit our implementation
+
+% Choose custom depth here:
+depth(5).
+
+% For simplicity, always assume player trying to maximize is red.
+maxPlayer(red).
+
+computerMove(Board, minimax, AvailableMoves, Colour, Move) :- 
+    depth(Depth),
+    Alpha is (-1 -Depth), % Alpha is best value that maximizer can guarantee
+    Beta is (1 + Depth), % Beta is best value that minimizer can guarantee
+    sortOrder([4,5,3,6,2,7,1], AvailableMoves, BestOrderAvailableMoves),
+    playMoves(BestOrderAvailableMoves, Depth, Alpha, Beta, Colour, Board, ScoreMoves),
+    nl,write('Minimax (score, move):'),write(ScoreMoves),nl,
+    getBestMove(Colour, ScoreMoves, Move).
+
+playMoves([], _, _, _, _, _, []).
+playMoves([Col|RestCols], Depth, Alpha, Beta, Colour, Board, [(Score, Col)|RestScoreMoves]) :-
+    playMove(Depth, Alpha, Beta, Colour, Board, Col, Score), % Try each move in available columns
+    (maxPlayer(Colour) ->
+        max(Alpha, Score, NewAlpha), % NewAlpha = max(Alpha, Score)
+        (NewAlpha > Beta ->
+            RestScoreMoves = []; % NewAlpha > Beta means we don't have to go deeper because maxPlayer will choose move pertaining to NewAlpha no matter what
+            playMoves(RestCols, Depth, NewAlpha, Beta, Colour, Board, RestScoreMoves))
+    ;
+        min(Beta, Score, NewBeta),
+        (NewBeta < Alpha ->
+            RestScoreMoves = [];
+        playMoves(RestCols, Depth, Alpha, NewBeta, Colour, Board, RestScoreMoves))).
+
+playMove(0, _, _, _, _, _, 0). % Reached max depth.
+playMove(Depth, Alpha, Beta, Colour, Board, Move, Score) :-
+    \+ Depth = 0,
+    newBoard(Board, Move, Colour, NewBoard),
+    (win(NewBoard) -> 
+    (maxPlayer(Colour) ->
+        Score is (1 + Depth) ; % Quicker wins are worth more
+        Score is (-1 - Depth)) ;
+        getAvailableMoves(NewBoard, NewAvailableMoves),
+        (NewAvailableMoves == [] ->
+            Score is 0;
```

Comment:
why is score 0 in this case? isn't it the outcome of getBestScore?

Reply:
In OOP, this translates to
```
if noMoreMovesAvailable (tie-game) {
  Score = 0 // tie
} else {
  NewDepth = Depth - 1
  playMoves(...) // Keep trying other moves
  getBestScore(..) // Get best score out of above moves
}
```

Reply:
oh i didn't see the if case gotcha